Core/include/Acts/TrackFinding/ComposableTrackStateCreator.hpp

```diff
@@ -0,0 +1,415 @@
+      const Logger& logger) const {
+    using PM = TrackStatePropMask;
+
+    using ResultTrackStateList =
+        Acts::Result<CkfTypes::BranchVector<TrackIndexType>>;
+    ResultTrackStateList resultTrackStateList{
+        CkfTypes::BranchVector<TrackIndexType>()};
+    CkfTypes::BranchVector<TrackIndexType>& trackStateList =
+        *resultTrackStateList;
+    trackStateList.reserve(end - begin);
+
+    std::optional<TrackStateProxy> firstTrackState{std::nullopt};
+    for (auto it = begin; it != end; ++it) {
+      auto& candidateTrackState = *it;
+
+      PM mask = PM::Predicted | PM::Filtered | PM::Jacobian | PM::Calibrated;
+      if (it != begin) {
+        // subsequent track states don't need storage for these as they will
+        // be shared
+        mask &= ~PM::Predicted & ~PM::Jacobian;
+      }
+      if (isOutlier) {
+        // outlier won't have separate filtered parameters
+        mask &= ~PM::Filtered;
+      }
+
+      // copy this trackstate into fitted states MultiTrajectory
+      auto trackState =
+          trackStates.makeTrackState(mask, candidateTrackState.previous());
+      ACTS_VERBOSE("Create SourceLink output track state #"
+                   << trackState.index() << " with mask: " << mask);
+
+      if (it != begin) {
+        // assign indices pointing to first track state
+        trackState.shareFrom(*firstTrackState, PM::Predicted);
+        trackState.shareFrom(*firstTrackState, PM::Jacobian);
+      } else {
+        firstTrackState = trackState;
+      }
+
+      // either copy ALL or everything except for predicted and jacobian
+      trackState.allocateCalibrated(candidateTrackState.calibratedSize());
+      trackState.copyFrom(candidateTrackState, mask, false);
+
+      auto typeFlags = trackState.typeFlags();
+      typeFlags.set(TrackStateFlag::ParameterFlag);
+      typeFlags.set(TrackStateFlag::MeasurementFlag);
+      if (trackState.referenceSurface().surfaceMaterial() != nullptr) {
+        typeFlags.set(TrackStateFlag::MaterialFlag);
+      }
+      if (isOutlier) {
+        // propagate information that this is an outlier state
+        ACTS_VERBOSE(
+            "Creating outlier track state with tip = " << trackState.index());
+        typeFlags.set(TrackStateFlag::OutlierFlag);
+      }
+
+      trackStateList.push_back(trackState.index());
+    }
+    return resultTrackStateList;
+  }
+
+  /// Default measurement selector which will return all measurements
+  /// @param candidates Measurement track state candidates
+  static Result<std::pair<typename std::vector<TrackStateProxy>::iterator,
+                          typename std::vector<TrackStateProxy>::iterator>>
+  voidMeasurementSelector(typename std::vector<TrackStateProxy>& candidates,
+                          bool& /*isOutlier*/, const Logger& /*logger*/) {
+    return std::pair{candidates.begin(), candidates.end()};
+  };
+};
+
+/// @brief Base class for the ComposableTrackStateCreator which simply delegates its tasks
```

Comment:
This is now derived, not base anymore, isn't it?

Reply:
exactly.  Thanks!